added post about Pidgin (#5)
benjamin-hodgson/benjamin-hodgson.github.io: posts/2018-01-26-announcing-pidgin-v2.0.md

* A type `Parser` which models a _parsing process_
* A collection of _primitive_ `Parser` objects which perform some simple individual parsing task
* A collection of _combinator_ functions which can build complex `Parser`s out of simpler ones

Taken together, we have an object model allowing you to compose parsers using an embedded domain-specific langauge. As a brief taste, here is a simple example parser which parses an identifier in a typical programming language.

```csharp
Parser<char, string> Identifier =
    Letter.Then(
        LetterOrDigit.ManyString(),
        (first, rest) => first + rest
    );

Assert.Equal("abc1", Identifier.ParseOrThrow("abc1"));
```

* `Identifier` is a `Parser<char, string>`, meaning it's a parsing process which consumes a sequence of `char`s and produces a `string`.
* `Letter` and `LetterOrDigit` are primitive parsers, each of which read a single character from the input string (ensuring that the character is alphabetical or alphanumeric, respectively).

> * `ManyString` is a combinator method which runs a parser repeatedly and packs the result into a `string` --- so `LetterOrDigit.ManyString()` is a parser which reads a sequence of alphanumeric characters.

Reviewer: I was following until the phrase "is a parser which reads a sequences of alphanumeric characters".  For some reason I don't follow that.

As I understand it, `LetterOrDigit` ensures that a character is a number or letter, and when the method ManyString() is called on it, what does it do? Does it checks multiple strings? Combine the strings? Just read them? Where do they go? Or does it do the callback function that's named after, which adds first + rest?

Author: ok so it sounds like i have a bit more explaining to do. I'll have a go here and if it makes sense I'll try to integrate it into the paragraph:

`LetterOrDigit` is a parser which _consumes_ a single character from the input stream, as long as that character is a letter or a digit. That means the input stream gets advanced to the next character if the current character is a letter or digit. If it's not, then the parser fails. The parser's return value is the character that it consumed.

`ManyString` is a combinator method which runs a parser in a loop until it fails. The `ManyString` parser's return value is all of the results from the loop, packed into a string.

So `LetterOrDigit.ManyString()` is a parser which consumes a sequence of letter-or-digit characters from the input stream, and returns that sequence of characters as a string.

Author: I'm thinking now that an introductory "how parsers work" article would be a good follow up post, rather than making this one even longer?

Reviewer: That explanation above is great - I get it much better now. An intro how parsers work article would be great too. Totally depends on who you imagine your audience to be. If they are people who already know how parsers work, which I'm assuming they probably are since they were looking for a parser combinator to use, then the additional explanation probably isn't necessary.

* `Then` is another combinator which runs two parsers in sequence and applies a function to the result. So, reading the parser as a whole, we can see that an `Identifier` consists of a single letter followed by a sequence of letters or digits.

> Parser combinators' power comes from their composability. The library comprises a small number of building blocks, which you can put together in rich and varied ways to build a parser which does what you need. The library's level of abstraction is a good fit for small-to-medium sized parsing tasks: it's not as high-level as a full-blown parser generator like Antlr, but it's much simpler to integrate. Rewriting our hand-written [JQL](https://www.benjamin.pizza/posts/2017-11-13-recursion-without-recursion.html) parser in Pidgin took around four times less code.

Reviewer: might be worth adding a link to definition of composability - I had to look it up

That's the overview --- I won't dive into a full tutorial on parser combinators because there are already [plenty](http://www.lihaoyi.com/post/EasyParsingwithParserCombinators.html) [of](http://webdoc.sub.gwdg.de/ebook/serien/ah/UU-CS/2008-044.pdf) [those](https://news.ycombinator.com/item?id=14600079) a mere Google away.

### How it works

> Pidgin's `Parser<TToken, T>` type represents a process which pulls `TToken`s one at a time from a (stateful) input stream, and either successfully returns a `T` or fails with an error message.

Reviewer: I don't know what a `TToken` is either 😑

Author: Here `TToken` is a type parameter (type parameters traditionally begin with `T`), meaning that `Parser` can consume a stream containing any type of token

```csharp
class ThenParser<TToken, T, U, R> : Parser<TToken, R>
{
    private readonly Parser<TToken, T> _first;
    private readonly Parser<TToken, U> _second;
    private readonly Func<T, U, R> _resultSelector;

    internal override Result<R> Parse(IParseState<TToken> input)
    {
        var result1 = _first.Parse(input);
        if (!result1.Success)
        {
            return Result.Error(result1.Error);
        }
        var result2 = _second.Parse(input);
        if (!result2.Success)
        {
            return Result.Error(result2.Error);
        }
        return Result.Success(_resultSelector(result1.Value, result2.Value));
    }
}
```

Error messages
--------------

Part of Pidgin's job is to report when you gave a parser bad input. For example, the top level of a C# file must contain `class`es and `namespace`s, so feeding a C# parser the text `Console.WriteLine("foo");` (without an enclosing `class`) should fail with some information about the problem:

```
Parse error.
    unexpected 'C'
    expected "class" or "namespace"
    at line 1, col 1
```

> A mistake I made early on in Pidgin's development nearly two years ago(!) was trying to pre-compute these content of these error messages. Under the assumption that parsers are typically built once and then run repeatedly, I wrote some code to examine your parser upon construction and try to predict the ways it could fail on unexpected input, in order to avoid constructing error messages at runtime. This code calculated a set of "expected" input strings, accounting for `Then`s and `Or`s by concatenating the strings in the set and by unioning the sets respectively. (The idea was that a parser like `Keyword("public").Optional().Then(Keyword("class"))` would report that it expected `"class" or "public class"`.)

Reviewer 2: ```suggestion
A mistake I made early on in Pidgin's development nearly two years ago(!) was trying to pre-compute the content of these error messages. Under the assumption that parsers are typically built once and then run repeatedly, I wrote some code to examine your parser upon construction and try to predict the ways it could fail on unexpected input, in order to avoid constructing error messages at runtime. This code calculated a set of "expected" input strings, accounting for `Then`s and `Or`s by concatenating the strings in the set and by unioning the sets respectively. (The idea was that a parser like `Keyword("public").Optional().Then(Keyword("class"))` would report that it expected `"class" or "public class"`.)
```

> This went catastrophically wrong for complex parsers. Here's an sketch of some code which parses left-associative mathematical operators with precedence, so `3^2 + 4 * 3^5 * 5` is parsed as `(3^2) + ((4 * (3^5)) * 5)`:

Reviewer 2: ```suggestion
This went catastrophically wrong for complex parsers. Here's a sketch of some code which parses left-associative mathematical operators with precedence, so `3^2 + 4 * 3^5 * 5` is parsed as `(3^2) + ((4 * (3^5)) * 5)`:
```

```csharp
var atom = Number;
var topPrecedence = atom.Then(Char('^').Then(atom).Many());
var midPrecedence = topPrecedence.Then(Char('*').Then(topPrecedence).Many());
var lowPrecedence = midPrecedence.SeparatedBy(Char('+').Then(midPrecedence).Many());
```

> `topPrecedence` expects a number, or a number followed by a sequence of exponents --- two possibilities. `midPrecedence` therefore expects a number, or a number followed by a sequence of exponents, or a number followed by a sequence of multiplications, or a number followed by a sequence of exponents followed by a sequence of multiplications --- four possibilities. `lowPrecedence` has eight possible expected inputs. The set of expected inputs blows up exponentially! Calculating the error messages in advance means you need to explore all of the (exponentially large number of) expected inputs in advance, which led to cosmologically long parser build times. And in any case, reporting a huge number of expected inputs does not make for a very good error message.

Reviewer: cosmologically is an A+ word. that's all.

Author: 🤓

I made some attempts to optimise this by using more efficient data structures, but I realised this was a losing proposition and decided to throw it out altogether in v2.0. Error messages are now computed at runtime, when the error actually occurs. This means I can be more precise about what the parser was expecting at that particular point in the input: once a parsing process has committed to a branch I can report expected inputs _from that branch_, rather than reporting all possible expected inputs.

> Implementing this efficiently was a challenge. Parse errors actually occur quite frequently in a parser combinator library, even in the happy path, because of the way the _prioritised choice_ operator `Or` works --- `String("foo").Or(String("bar"))` only tries `bar` if an attempt to parse `foo` failed. So I tried to implement this change without allocating heap memory every time a parser fails. When a parser fails, it saved its expected inputs in a stack implemented on top of pooled memory (using `ArrayPool`), which are then popped if the error gets discarded. (The way that certain parsers manipulate their children's error messages adds some interesting complications here, which I've described in a [long comment](https://github.com/benjamin-hodgson/Pidgin/blob/60c7734393719d11714158b201c99976ec48ffb9/Pidgin/ParseState.Error.cs#L36-L74).)

Reviewer: think should be "When a parser fails, it *saves* its expected inputs?

Author: good catch, thank you

`Span`
------

> > **Note**: You can watch [my esteemed colleague Gervasio](https://twitter.com/g3rv4) and me carrying out the work I describe here in [a live-stream on YouTube](https://www.youtube.com/watch?v=O23OLkQtiS4). It was pretty fun!

Reviewer 2: ```suggestion
> **Note**: You can watch [my esteemed colleague Gervasio](https://twitter.com/g3rv4) and I carrying out the work described here in [a live-stream on YouTube](https://www.youtube.com/watch?v=O23OLkQtiS4). It was pretty fun!
```

Author: "me" is grammatically correct here. Imagine if Gervasio wasn't involved, the sentence would say "you can watch me carry out the work"

Reviewer 2: I dunno, I barely even speak the language

Author: more fluent in C# than English

Assorted Performance Improvements
---------------------------------

Performance has always been one of Pidgin's priorities --- I'm proud that Pidgin is C#'s fastest parser combinator library (that I know of!) --- but there's always room for improvement. In my tests Pidgin still runs somewhat slower than F#'s FParsec library, for example.) In this release I made some architectural changes to `IParseState` to help close that gap.

### Buffering Uniformly

As I mentioned earlier, Pidgin has several internal implementations of `IParseState`, each of which implements a streaming abstraction on top of a different type of input. Parsers may need to _backtrack_ on failure (using the `Try` combinator), so you can't always discard a token as soon as you've seen it. Some `IParseState` implementations --- specifically the ones that are built on top of streaming storage (like `Stream`) --- therefore _buffer_ their input into an array. The ones that are backed by in-memory storage like `string` don't need to buffer because their data is already in memory.

> I decided to move the buffering logic to a shared part of the code. Now _all_ `IParseState` implementations buffer thier input, even the in-memory ones. On its own this should make the code slower (it's more expensive to copy a `string` into an array than not to!), but it enables all of the optimisations I'm about to describe.

Reviewer 2: ```suggestion
I decided to move the buffering logic to a shared part of the code. Now _all_ `IParseState` implementations buffer their input, even the in-memory ones. On its own this should make the code slower (it's more expensive to copy a `string` into an array than not to!), but it enables all of the optimisations I'm about to describe.
```

```csharp
struct ParseState<TToken>
{
    private TToken[] _buffer;
    private int _currentPosition;

    public bool HasCurrent => _currentPos >= 0 && _currentPos < _buffer.Length;
    public TToken Current => _buffer[_currentPosition];
}
```

These properties are short and non-virtual, so they are good inlining candidates. After inlining them, the JIT sees simple array-manipulating code, which it can easily optimise in the context of the containing method. This simple change resulted in a 15-20% performance improvement across the board.

### Chunking

The `Stream` version of `Advance` called `stream.Read()`, which reads and returns a single byte from the stream. This is not as efficient as `stream.Read(byte[])`, which reads a chunk of bytes from the stream and copies them into an array. (The former requires _n_ virtual method calls to read _n_ bytes, whereas the latter requires only _n / `array.Length`_.)

I replaced `ITokenStream`'s `Advance` and `Peek` methods with a method which reads a chunk of tokens into the `ParseState`'s buffer.

```csharp
interface ITokenStream<TToken>
{
    void ReadInto(TToken[] buffer, int start, int count);
}
```

> Why not design this signature as `void ReadInto(Span<TToken> span)`? I'd love to! Sadly that would preclude some implementations of `ITokenStream`. Methods like `stream.Read(Span<byte> span)` are available only on .NET Core, and as far as I'm aware Microsoft has no plans to backport them, so for compatibility with the desktop framework I'm stuck using arrays. This is irksome, as [I've said before](/posts/2018-12-06-zooming-in-on-field-accessors.html) --- what's the point of designing a feature for library authors if you're not going to support it properly? It's obvious that the desktop framework is reaching the end of its life. Microsoft should just grasp the nettle and admit it.

Reviewer 2: That last bit is speculation, not sure I'd bother ;)

```csharp
public ReadOnlySpan<TToken> Peek(int count);
```

Using `Peek`, a parser like `String("foo")` can now look at three characters from the input to see if they match `foo`, rather than one at a time. This runs about 20% faster for long strings.

Make Your Own Opportunities
---------------------------

Indulge me for a moment while I dispense some unsolicited career advice. For a long time, performance engineering was something that [other](https://twitter.com/marcgravell/), [cleverer](https://twitter.com/davidfowl) [people](https://mattwarren.org/) did while I observed from a distance with awe. Why is ASP.NET fast? Because the guys that wrote it are Actual Wizards.

I've been lucky enough to work with some of those clever people for a few years now, and actually they're not Actual Wizards. They just have more experience than me --- they know how to use and interpret profiles and benchmarks, they've seen enough to know what does and doesn't work, and they've had practice coming up with ideas to improve performance.

> The good news is that it's never too late or too early to start building your own experience level, no matter where you are in your career. For me that meant setting myself the goal of making my open source libraries as fast as I could. Working on performance-focused libraries has been hugely rewarding for me, and learning how to write low-level C# has paid off at work. For you, it could mean picking a slow part of your codebase at work and trying to make it faster, or pair-programming with a colleague, or thoroughly reading through an existing highly optimised codebase. Making your own opportunities to learn new things is the best way to get better.

Reviewer 2: +1 might be worth mentioning that a lot of performance work doesn't even need to be this low-level. Being aware of where you're doing IO (DB calls, caching, etc.) and an understanding of different data structures can make a huge difference!

Author: Good idea, thanks!

> The good news is that performance optimisation usually doesn't involve fiddly low-level programming like this --- simply being aware of when your code performs IO makes a big difference --- and it's never too early or late to start building your experience level, no matter where you are in your career. For me, that meant setting myself the goal of making my open source libraries as fast as I could. For you, it could mean picking a slow part of your codebase at work and trying to make it faster, or pair-programming with a colleague, or thoroughly reading through some unfamiliar code. Making your own opportunities to learn new things is the best way to get better.

Reviewer: I love this sentiment <3

---
title: Announcing Pidgin v2.0
subtitle: Errors, `Span`, and performance
---

> Happy birthday to me! I'm pleased to announce that version 2.0 of my functional parsing library, [Pidgin](https://github.com/benjamin-hodgson/Pidgin), is now available [on Nuget](https://www.nuget.org/packages/Pidgin). In this release I've focused on error messages, performance, and `Span` support.

Reviewer 3: Happy Birthday! 🎂

Author: ❤️
